language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/LFUtil.scala

```diff
@@ -266,6 +266,16 @@ final case class LFUtil(
       case _ => false
     }
   }
+
+  protected[this] override def precacheVariance(interface: Interface) = {
+    import UsedTypeParams.ResolvedVariance
+    val resolved = interface.typeDecls.foldLeft(ResolvedVariance.Empty) {
+      case (resolved, (id, InterfaceType.Template(_, _))) =>
+        resolved.allCovariantVars(id, interface)._1
```

Comment:
I don’t understand why we can only recurse on templates. Doesn’t the same problem apply to choice types?

Reply:
We only need to reach for the "roots" here; `allCovariantVars` itself will recur into referenced datatypes. Nor do we need to be perfect here; if we missed any types, they'll just be resolved later when we ask them to be resolved. We are only looking to precache the variances we're actually going to need, while not resolving tons of variances we won't need.

Reply:
But the record template type doesn’t need to reference choice argument or return types. What makes sure we hit them?

Reply:
_Oh_, I read "choice types" as "variant types" and just reacted internally "well that shouldn't make a difference". It's little problem to add choice and key types here, though it's little harm to forget them either.

Reply:
Why is there no harm in forgetting them? Doesn’t that mean we get variance of those types wrong?

Reply:
(As discussed out-of-band) it's a precache, so we just compute later if we miss it. Anyway, 53940cd3696 adds choice and key types to the precache.